crates/ruff_linter/src/rules/ruff/rules/for_variable_used_after_block.rs

```diff
@@ -0,0 +1,109 @@
+    scope: &Scope,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    // Keep track of the node_ids of variable references that have already been
+    // accounted for as part of a different variable binding. This helps us avoid
+    // complaining when people use the same variable name in multiple blocks under the
+    // same scope.
+    let mut known_good_reference_node_ids: Vec<NodeId> = Vec::new();
+
+    let all_bindings: Vec<(&str, BindingId)> = scope.all_bindings().collect();
+    // We want to reverse the bindings so that we iterate in source order and shadowed
+    // bindings come first. This way we can gather `known_good_reference_node_ids` and
+    // mark off things as we go which will allow us to ignore later bindings trying to
+    // reference the same variable.
+    let reversed_bindings = all_bindings.iter().rev();
+
+    // Find for-loop variable bindings
+    let loop_var_bindings = reversed_bindings
+        .map(|(name, binding_id)| (name, checker.semantic().binding(*binding_id)))
+        .filter_map(|(name, binding)| binding.kind.is_loop_var().then_some((name, binding)));
+
+    for (&name, binding) in loop_var_bindings {
+        let binding_statement = binding.statement(checker.semantic()).unwrap();
+        let binding_source_node_id = binding.source.unwrap();
+        // The node_id of the for-loop that contains the binding
+        let binding_statement_id = checker.semantic().statement_id(binding_source_node_id);
+
+        // Loop over the references of those bindings to see if they're in the same block-scope
+        'references: for reference in binding.references() {
+            let reference = checker.semantic().reference(reference);
+            let reference_node_id = reference.expression_id().unwrap();
+
+            // Skip any reference that come before the control var binding in the source
+            // order, skip it because people can assign and use the same variable name
+            // above the block.
+            if reference.range().end() < binding_statement.range().start() {
+                continue;
+            }
+
+            // Traverse the hierarchy and look for a block match
+            let statement_hierarchy = checker.semantic().parent_statement_ids(reference_node_id);
+
+            for ancestor_node_id in statement_hierarchy {
+                if binding_statement_id == ancestor_node_id {
+                    known_good_reference_node_ids.push(reference_node_id);
+                    continue 'references;
+                }
+            }
+
+            // If the reference wasn't used in the same block, report a violation/diagnostic
+            if !known_good_reference_node_ids.contains(&reference_node_id) {
+                diagnostics.push(Diagnostic::new(
+                    ForVariableUsedAfterBlock {
+                        control_var_name: name.to_owned(),
+                    },
+                    reference.range(),
+                ));
+            }
+        }
```

Comment:
When I remove the `known_good_reference_node_ids` check, then no test fails. Can we add a test that demonstrates why the check is necesseary or remove the check?
```suggestion
            diagnostics.push(Diagnostic::new(
                ForVariableUsedAfterBlock {
                    control_var_name: name.to_owned(),
                },
                reference.range(),
            ));
        }
```

Reply:
Good call on this :thumbsup: The `known_good_reference_node_ids` evolved first to [solve the shadowed binding cases](https://github.com/astral-sh/ruff/commit/eaf2a35753b8bb3777d455be8e336a3072887f16) but my [fix later on to allow previous references](https://github.com/astral-sh/ruff/commit/a7de1cb892d3ca34b35fb90daeae2367465e99b4) also ended up covering the shadowed cases without me realizing (much more elegant too). Sorry for not noticing :grimacing:

Removed the `known_good_reference_node_ids` logic :white_check_mark: